src/main/ruby/truffleruby/core/splitter.rb

```diff
@@ -36,157 +36,219 @@
-        ret << last_match.post_match
-      elsif ret.empty?
-        ret << string.dup
-      end
+      def split_type_regexp(string, pattern, limit, &block)
+        start = 0
+        ret = []
+        count = 0
+        empty_count = 0
+        limited = limit > 0
+
+        last_match = nil
+        last_match_end = 0
+
+        while match = Truffle::RegexpOperations.match(pattern, string, start)
+          break if limited && limit - count <= 1
+
+          collapsed = Truffle::RegexpOperations.collapsing?(match)
+
+          unless collapsed && (match.byte_begin(0) == last_match_end)
+            substring = Truffle::RegexpOperations.pre_match_from(match, last_match_end)
+            empty_count = add_substring(string, ret, substring, empty_count, &block)
+
+            # length > 1 means there are captures
+            if match.length > 1
+              match.captures.compact.each do |capture|
+                empty_count = add_substring(string, ret, capture, empty_count, &block)
+              end
+            end
+
+            count += 1
+          end
+
+          start = match.byte_end(0)
+          if collapsed
+            start += 1
+          end
 
-      # Trim from end
-      if Primitive.undefined?(limit) || limit == 0
-        while s = ret.at(-1) and s.empty?
-          ret.pop
+          last_match = match
+          last_match_end = last_match.byte_end(0)
         end
-      end
 
-      ret
-    end
+        if last_match
+          empty_count = add_substring(string, ret, last_match.post_match, empty_count, &block)
+        elsif ret.empty?
+          empty_count = add_substring(string, ret, string.dup, empty_count, &block)
+        end
 
-    def self.split_on_string(string, pattern, trim_end)
-      pos = 0
+        if tail_empty?(limit)
+          add_empty(string, ret, empty_count, &block)
+        end
+
+        result(string, ret, &block)
+      end
 
-      ret = []
 
-      pat_size = pattern.bytesize
-      str_size = string.bytesize
+      def add_substring(string, array, substring, empty_count, &block)
+        return empty_count + 1 if substring.length == 0 # remember another one empty match
 
-      while pos < str_size
-        nxt = Primitive.find_string(string, pattern, pos)
-        break unless nxt
+        add_empty(string, array, empty_count, &block)
 
-        match_size = nxt - pos
-        ret << string.byteslice(pos, match_size)
+        add_or_call(array, substring, &block)
 
-        pos = nxt + pat_size
+        0 # always release all empties when we get non empty substring
       end
 
-      # No more separators, but we need to grab the last part still.
-      ret << string.byteslice(pos, str_size - pos)
+      def add_empty(string, array, count, &block)
+        count.times { add_or_call(array, empty_string(string), &block) }
+      end
 
-      if trim_end
-        while s = ret.at(-1) and s.empty?
-          ret.pop
+      def add_or_call(array, element, &block)
+        if block_given?
```

Comment:
Minor but I'd use `if block` here. In TruffleRuby `if block` vs `if block_given?` (and not having a variable for the block) has no performance difference once compiled. In interpreter, `if block_given?` needs to do a method lookup once to make sure it's `Kernel#block_given?` (and relies on AST-inlining otherwise it's expensive as it needs access to the caller frame).

Reply:
Good point, done :+1:

Regarding `result` inlining - @bjfish has already suggested this before, so 2vs1, is done too :)